include/nbl/builtin/hlsl/bxdf/common.hlsl

```diff
@@ -48,358 +139,976 @@ struct Basic
+    ((NBL_CONCEPT_REQ_TYPE_ALIAS_CONCEPT)(surface_interactions::Anisotropic, typename T::anisotropic_type))
+);
+#undef param
+#undef aniso
+#undef iso
+#undef _sample
+#undef pdf
+#undef spec
+#undef bxdf
+#include <nbl/builtin/hlsl/concepts/__end.hlsl>
+
+#define NBL_CONCEPT_NAME MicrofacetBxDF
+#define NBL_CONCEPT_TPLT_PRM_KINDS (typename)
+#define NBL_CONCEPT_TPLT_PRM_NAMES (T)
+#define NBL_CONCEPT_PARAM_0 (bxdf, T)
+#define NBL_CONCEPT_PARAM_1 (spec, typename T::spectral_type)
+#define NBL_CONCEPT_PARAM_2 (pdf, typename T::scalar_type)
+#define NBL_CONCEPT_PARAM_3 (_sample, typename T::sample_type)
+#define NBL_CONCEPT_PARAM_4 (iso, typename T::isotropic_type)
+#define NBL_CONCEPT_PARAM_5 (aniso, typename T::anisotropic_type)
+#define NBL_CONCEPT_PARAM_6 (isocache, typename T::isocache_type)
+#define NBL_CONCEPT_PARAM_7 (anisocache, typename T::anisocache_type)
+#define NBL_CONCEPT_PARAM_8 (param, typename T::params_t)
+NBL_CONCEPT_BEGIN(9)
+#define bxdf NBL_CONCEPT_PARAM_T NBL_CONCEPT_PARAM_0
+#define spec NBL_CONCEPT_PARAM_T NBL_CONCEPT_PARAM_1
+#define pdf NBL_CONCEPT_PARAM_T NBL_CONCEPT_PARAM_2
+#define _sample NBL_CONCEPT_PARAM_T NBL_CONCEPT_PARAM_3
+#define iso NBL_CONCEPT_PARAM_T NBL_CONCEPT_PARAM_4
+#define aniso NBL_CONCEPT_PARAM_T NBL_CONCEPT_PARAM_5
+#define isocache NBL_CONCEPT_PARAM_T NBL_CONCEPT_PARAM_6
+#define anisocache NBL_CONCEPT_PARAM_T NBL_CONCEPT_PARAM_7
+#define param NBL_CONCEPT_PARAM_T NBL_CONCEPT_PARAM_8
+NBL_CONCEPT_END(
+    ((NBL_CONCEPT_REQ_TYPE)(T::scalar_type))
+    ((NBL_CONCEPT_REQ_TYPE)(T::isotropic_type))
+    ((NBL_CONCEPT_REQ_TYPE)(T::anisotropic_type))
+    ((NBL_CONCEPT_REQ_TYPE)(T::sample_type))
+    ((NBL_CONCEPT_REQ_TYPE)(T::spectral_type))
+    ((NBL_CONCEPT_REQ_TYPE)(T::quotient_pdf_type))
+    ((NBL_CONCEPT_REQ_TYPE)(T::isocache_type))
+    ((NBL_CONCEPT_REQ_TYPE)(T::anisocache_type))
+    ((NBL_CONCEPT_REQ_EXPR_RET_TYPE)((bxdf.eval(param)), ::nbl::hlsl::is_same_v, T::spectral_type))
+    ((NBL_CONCEPT_REQ_EXPR_RET_TYPE)((bxdf.generate(aniso,aniso.N,anisocache)), ::nbl::hlsl::is_same_v, typename T::sample_type))
+    //((NBL_CONCEPT_REQ_EXPR_RET_TYPE)((bxdf.template pdf<LS,I>(_sample,iso)), ::nbl::hlsl::is_scalar_v))
+    ((NBL_CONCEPT_REQ_EXPR_RET_TYPE)((bxdf.quotient_and_pdf(param)), ::nbl::hlsl::is_same_v, typename T::quotient_pdf_type))
+    ((NBL_CONCEPT_REQ_TYPE_ALIAS_CONCEPT)(Sample, typename T::sample_type))
+    ((NBL_CONCEPT_REQ_TYPE_ALIAS_CONCEPT)(spectral_of, typename T::spectral_type, typename T::scalar_type))
+    ((NBL_CONCEPT_REQ_TYPE_ALIAS_CONCEPT)(IsotropicMicrofacetCache, typename T::isocache_type))
+    ((NBL_CONCEPT_REQ_TYPE_ALIAS_CONCEPT)(AnisotropicMicrofacetCache, typename T::anisocache_type))
+);
+#undef param
+#undef anisocache
+#undef isocache
+#undef aniso
+#undef iso
+#undef _sample
+#undef pdf
+#undef spec
+#undef bxdf
+#include <nbl/builtin/hlsl/concepts/__end.hlsl>
```

Comment:
let is be `BxDF<T>` and then just add constraints that `generate`,`eval` and `quotient_and_pdf` have the overloads which take microfacet caches

Reply:
So then will these bxdfs need to have `generate` (`eval` and `quotient_and_pdf` take in param structs) that don't take in a microfacet cache? I think MicrofacetBxdf is also not the name I'm looking for. But these would be the ggx and beckmann bxdfs (both brdf and bsdf).

Reply:
contnued in https://github.com/Devsh-Graphics-Programming/Nabla/pull/899/files?diff=unified&w=1#r2191052285